### DISCUSSION TITLE
Restic Restore - restic-wait: Not Found Error

### DISCUSSION THREAD

**Discussion by dcplaya** [vmware-tanzu/velero#3385] [Community support Q&A]
I recently restored a large PV from a restic backup (~250GB, many many small files). The restore completed successfully, but during the restore I noticed the velero-restic-helper init container would report the following errors over and over again.

(Sorry for the picture, its all I have without typing it all out by hand. Ill leave 1 typed error so people can search for it)
The errors showed up consistently for the entire restore process, I cannot remember if the path name changed during the restore (it took hours, I didn't monitor it the entire time)

The restore seems successful, the application started up with no glaring issues. I am just looking for info on what this error means, what causes it, do I need to worry about it, ect.

`restic-wait Not Found: /restores/config/.velero/3c3819c6-f522-41212-b64c-c7672489d423`

![image](https://user-images.githubusercontent.com/1105320/106385763-c4981200-6386-11eb-9577-f7e73d62bcf3.png)

**Comment by dsu-igeek** [comment]
**The output of the following commands will help us better understand what's going on**:
(Pasting long output into a [GitHub gist](https://gist.github.com) or other pastebin is fine.)

- `kubectl logs deployment/velero -n velero`
- `velero backup describe <backupname>` or `kubectl get backup/<backupname> -n velero -o yaml`
- `velero backup logs <backupname>`
- `velero restore describe <restorename>` or `kubectl get restore/<restorename> -n velero -o yaml`
- `velero restore logs <restorename>`


**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]


**Environment:**

- Velero version (use `velero version`):
- Velero features (use `velero client config get features`):
- Kubernetes version (use `kubectl version`):
- Kubernetes installer & version:
- Cloud provider or hardware configuration:
- OS (e.g. from `/etc/os-release`):

**Reply by tico24** [reply]
I'm afraid I don't have those logs to hand (yet?), but I can add a little more information:

Using velero 1.5.2, it works as expected, the output from the init container is as follows:
```
 Not found: /restores/storage/.velero/0d75938c-d33c-40d9-977c-a5073cebc25d
 Not found: /restores/storage/.velero/0d75938c-d33c-40d9-977c-a5073cebc25d
 Found /restores/storage/.velero/0d75938c-d33c-40d9-977c-a5073cebc25dAll restic restores are done
 stream closed
```

Using velero 1.5.3, it fails, the `Not Found` lines repeat indefinitely. 

To clarify a little about what @dcplaya wrote. What I'm seeing is:

1. The restore is marked as 'completed'
2. The PV is restored, (and the pod and the namespace etc)
3. The Pod starts up, k8s recognises that the PV is there.
4. The init container 'restic-wait' then starts churning out the errors above. As it's an init container, it then blocks the pod from properly starting.

If I delete the restored pod, and let it re-spawn itslef, the startup process is then as expected, the init container finds the PV and the pod starts.

So I'm guessing there's a newly added race-condition in the restic-wait container that prevents it from properly seeing the PV?


The above tests are on a 1GB PV containing just one text file with the word 'TEST' in it, so it's not a  volume size issue. I spotted the problem because I have a nightly restore test running and it has been failing since I upgraded to the 1.5.3 containers.

**Reply by zubron** [reply]
@tico24 Just to confirm, in Velero 1.5.3, the restic pod volume restore does not complete and you no longer see the `All restic restores are done` message in the logs?

You can see the changelogs for the v1.5.3 release here: https://github.com/vmware-tanzu/velero/releases/tag/v1.5.3
The change in behaviour that you are seeing may be related to this change: https://github.com/vmware-tanzu/velero/pull/3011
Do any of the pods that you are restoring with restic have other init containers?

**Reply by tico24** [reply]
@zubron  Your confirmation is correct. 

My pod does not have any init containers. It's simply the latest nginx pod with the pv mounted at the html directory. 

The pv just contains an index.html containing just the word 'TEST'.

**Reply by zubron** [reply]
This will require more investigation on our part. Can you provide the following logs:

- `kubectl logs deployment/velero -n velero`
- `velero restore describe <restorename>` or `kubectl get restore/<restorename> -n velero -o yaml`
- `velero restore logs <restorename>`
- `kubectl get podvolumerestores.velero.io -n velero -o yaml`

**Reply by tico24** [reply]
I'll have to upgrade again to 1.5.3 so this will take time. 

I can say that those logs won't tell you anything, they all look normal.

**Reply by zubron** [reply]
That may be the case, but it would help to know the details of the restore and the volumes being restored so we can verify that the restic init container is looking for the correct done file to know that the volume has finished restoring.

**Comment by JarnoRFB** [comment]
I am getting the same error when trying to restore into a different namespace. The kubernetes resources are still created. It is really strange, since restoring into the same namespace works perfectly fine.

I use a command like 

`velero restore create --from-backup test-backup3 --namespace-mappings mlflow-feature-hotfix:something-else`

Note that mlflow-feature-hotfix ist just a sample application with a postgres that I use to have some state while verifying the backups.

I have velero `1.5.3` installed via helm with the following values.

```yaml
initContainers:
  - name: velero-plugin-for-aws
    image: velero/velero-plugin-for-aws:v1.1.0
    imagePullPolicy: IfNotPresent
    volumeMounts:
      - mountPath: /target
        name: plugins
configuration:
  provider: aws
  backupStorageLocation:
    bucket: ${bucket}
    prefix: velero
  volumeSnapshotLocation:
    name: aws-default
  defaultVolumesToRestic: true
deployRestic: true
credentials:
  name: velero-aws
  secretContents:
    cloud: |
      [default]
      aws_access_key_id=${aws_access_key_id}
      aws_secret_access_key=${aws_secret_access_key}

```

I am runnning on Rancher on premise. I put the logs files here https://gist.github.com/JarnoRFB/25b6343e45ca63945ebc9eca457774ab , but to me the logs of the successful backup to the same namespace looked almost the same as the logs of the failing backup to the different namespace.

Any help much appreciated!

**Reply by tico24** [reply]
I forgot to mention that I am also restoring to a new namespace.

**Reply by JarnoRFB** [reply]
@tico24 that is interesting, as this seems be the root of the problem. Have you tried restoring to the same namespace also?

**Reply by tico24** [reply]
@JarnoRFB I'm afraid not. If I get a chance, I will try today.

**Reply by sevenval-admins** [reply]
We are seeing the same problem. A restore to the same namespace works. A mapping into another namespace does not work.

As an experiment, I downgraded to Velero 1.5.2 and was able to successfully do the restore, which did not work with 1.5.3.

Environment:

Kubernetes: v1.18.8
Storage: Netapp
CSI driver: Trident 20.10.1

**Reply by JarnoRFB** [reply]
@sevenval-admins Same for me.

**Comment by zubron** [comment]
@dcplaya As part of restoring volumes using restic, Velero injects an init container in the pod being restored that watches for a "done file" which indicates that the volume has been fully restored by restic. This done file is written by a restic pod in the velero namespace on the node where your pod is being restored. The message `Not Found: /restores/config/.velero/3c3819c6-f522-41212-b64c-c7672489d423` from the restic-wait init container is a log message indicating that the done file has not yet been written and that the volume is not fully restored yet. This log message is printed every second.

Given that you were restoring a large amount of data, this may take some time and so you would see many entries of this log message. The restore should eventually complete though, and you should see `All restic restores are done` as the final entry. If your pod and volume are restored, and you see the `All restic restores are done` in the init container logs then there is no cause for concern - this is expected behaviour.

**Reply by tico24** [reply]
@zubron I have already proved that this is definitely not the case. A 5kb file hangs indefinitely (I left it over 48hours). Manually deleting the stuck pod resolves the issue. As soon as the new pod spawns, the checker init container works.

**Reply by zubron** [reply]
I was responding to @dcplaya's original message, which included a question about whether these messages were expected. I have responded to your follow up inline.

**Reply by JarnoRFB** [reply]
@zubron Good to know that theses log message are expected! Though I dare say that they are a bit confusing and do not really confer that there is still work going on. 

However, in my case, as well as it seems to be for @tico24, the restore never finishes and times out eventually. Also for me, the same backup can be perfectly restored into the same namespace, which indicates, that it is not a problem with the size of the persistent volume.

**Reply by dcplaya** [reply]
I can confirm that I had no issues (except for the `NOT FOUND` "error", maybe error should be more descriptive?) restoring the PV to the same namespace.

I have tried restoring the PV to a different namespace and I am unable to, I see similar results as @JarnoRFB and @tico24

**Reply by zubron** [reply]
Thanks for the confirmation everyone! I'll open an issue to investigate the restore to different namespaces. In the meantime, if you can, I'd recommend using 1.5.2 as the issue doesn't seem to be present in that release.

**Reply by JarnoRFB** [reply]
Just looking at the changelog, it could be related to https://github.com/vmware-tanzu/velero/pull/3051, since the namespace is now used to identify the pod volume backup.

**Comment by zubron** [comment]
Thanks everyone for your input here regarding the issue when restoring using a namespace mapping. It's definitely a bug and we've opened #3467 to address this. It will be included in the v1.6 release.

**Comment by hervedevos** [comment]
I am able to reproduce the same "Not found" error on restic-wait side car with Velero 1.8.1 version + restic persistent volume backup but in a bit different use case:
- Backup namespace "bitnami-pg" from one GKE cluster 
- Restore in same "bitnami-pg"  name to another GKE cluster

Find attached logs of restore :
[restores_bitnami-pg-2-20220421161827_restore-bitnami-pg-2-20220421161827-results.log](https://github.com/vmware-tanzu/velero/files/8531793/restores_bitnami-pg-2-20220421161827_restore-bitnami-pg-2-20220421161827-results.log)
[restores_bitnami-pg-2-20220421161827_restore-bitnami-pg-2-20220421161827-logs.log](https://github.com/vmware-tanzu/velero/files/8531799/restores_bitnami-pg-2-20220421161827_restore-bitnami-pg-2-20220421161827-logs.log)
[restic-wait.log](https://github.com/vmware-tanzu/velero/files/8531804/restic-wait.log)

As soon as I delete the waiting pod it restarts without any issue (and with data fully restored)
